[EthFlow]#1294 handle failed txs

alfetopito opened the pull request:
# Summary

Part of #1294 

Handle creation transaction failure

<img width="815" alt="Screenshot 2022-12-20 at 14 43 43" src="https://user-images.githubusercontent.com/43217/208699072-e3782d46-02be-469b-b9b7-5858c3594e5e.png">

![Screenshot 2022-12-20 at 15 27 19](https://user-images.githubusercontent.com/43217/208703464-142bca0d-d87e-4388-ae14-4cd9e722972a.png)


- New order state `invalid` - this will be a new state returned by the backend
- Updated stepper to display cancellation failure

## Notes
1. Not tested with transaction cancellation/speed up. May or may not work (more likely won't)
2. Cancellation transaction failure not yet handled

  # To Test

1. Place an ETH sell order for 1 ETH
* Tx will fail
* Stepper will be updated to show the status like the picture above
2. Place regular ETH sell orders
* Should work as usual

How do you fail a tx, you ask?
I've made a temporary change so every ETH flow order will always fail when value is exactly `1 ETH` in this PR :)

alfetopito commented on src/custom/state/enhancedTransactions/updater/FinalizeTxUpdater.tsx:
```diff
@@ -69,12 +69,6 @@ function finalizeEthereumTransaction(
   const { chainId, addPopup, dispatch } = params
   const { hash } = transaction
 
-  const ethFlowInfo = transaction.ethFlow
-  if (ethFlowInfo) {
-    params.removeInFlightOrderId(ethFlowInfo.orderId)
-    setTimeout(() => params.removeInFlightOrderId(ethFlowInfo.orderId), DELAY_REMOVAL_ETH_FLOW_ORDER_ID_MILLISECONDS)
-  }
-
```
Moved this down with the existing conditional

Also removed line `74` as it makes the timeout irrelevant
Or did I read this wrong?

@anxolin

anxolin replied:
🙈 I think i had an issue rebasing, where I probably click on keep both. Good catch

nenadV91 commented on src/cow-react/modules/swap/pure/EthFlow/EthFlowStepper/steps/Progress2.tsx:
```diff
@@ -11,40 +11,26 @@ export function Progress2({ order, refund, cancellation }: EthFlowStepperProps)
+    const isInvalid = state === SmartOrderStatus.INVALID
+
+    if (isInvalid) {
+      return { status: 'error', value: 0 }
```
Should these status values be defined in a enum?

alfetopito replied:
It's defined [here](https://github.com/cowprotocol/cowswap/blob/bf11c21cddc19dd86263e283d6fecd409c6f323e/src/cow-react/modules/swap/pure/EthFlow/EthFlowStepper/StatusIcon.tsx#L6) and I think this is fine.

nenadV91 commented on src/cow-react/modules/swap/pure/EthFlow/EthFlowStepper/steps/Step1.tsx:
```diff
@@ -2,13 +2,15 @@ import React from 'react'
   const isCreating = state === SmartOrderStatus.CREATING
+  const isInvalid = state === SmartOrderStatus.INVALID
 
   let label: string, stepState: StatusIconState, icon: string
```
I thought we said that we won't use `let` anymore.
But I'm fine with this.

alfetopito replied:
Well, you complain with Anxo who first wrote this :D

I could refactor it but out of scope of the current PR.

What we can do is to make sure we don't as much as possible and remember to point this out when introduced

anxolin replied:
Yep, my fault. and yep, not the end of the world in my opinion this has literally 10 lines of code, and TS ensures that if branches set a value to the 3 variables. 

I think I avoid using it for the others, like `Progress2`. Agree with trying to avoid it

nenadV91 commented on src/cow-react/modules/swap/pure/EthFlow/EthFlowStepper/steps/Step2.tsx:
```diff
@@ -71,12 +72,12 @@ export function Step2({ order, cancellation }: EthFlowStepperProps) {
       state: 'success',
```
These state values could also be an enum then

nenadV91 reviewed (approved):
Besides some minor comments, codewise LGTM :+1:

anxolin reviewed (approved):
🚀 Great PR!

One red flag we have in our project is how many changes are required to add a new state. But I think is good call how you handle this 👍

anxolin commented:
@alfetopito i have some questions, but I tried to add them in form of a PR https://github.com/cowprotocol/cowswap/pull/1776

elena-zh commented:
Hey @alfetopito , great job!
Some minor issues I faced: 
1. It would be nice to cross out `Receive <token>` on the 3rd step
![cross out](https://user-images.githubusercontent.com/70885163/208754002-6f52eb52-96df-41fc-8fa1-f14b8ee20af6.png)
2. There is no any failed order pop-up (and no sound) when TX fails. Can we add it? 

Then, I have retested changes for a TX cancellation/speed up. 
Cancellation does not work: the order hangs in 'Creating' state
![does not](https://user-images.githubusercontent.com/70885163/208754257-950b7300-6321-4bb6-9db0-2a76a5a9f854.png)

As for speed up, the app does not detect a replaced TX hash and navigates a user to a not existing hash
![image](https://user-images.githubusercontent.com/70885163/208754417-a3183803-a427-4c00-9a3e-b111ec5415aa.png)
But the order is created, and the stepper shows correct steps for the order status. 

Thanks

alfetopito commented:
@elena-zh addressed 1
2... might address in another PR. Can you create an issue for it?

@anxolin I've refactored the stepper as mentioned on Slack and updated cosmos adding the missing cases, so https://github.com/cowprotocol/cowswap/pull/1776 is outdated
Take a look when you get back :)

anxolin commented on src/cow-react/modules/swap/pure/EthFlow/EthFlowStepper/index.cosmos.tsx:
```diff
@@ -350,6 +313,47 @@ const STEPS: Step[] = [
+        rejectedReason: 'Transaction failed',
+      },
+      creation: { ...defaultProps.creation, failed: true },
+    },
```
Adding pic for this. In case it helps other reviewers

<img width="1116" alt="image" src="https://user-images.githubusercontent.com/2352112/209008387-a4f9343f-045f-4f2f-bc0d-08e4ec2387c7.png">


It looks good to me!

anxolin commented on src/cow-react/modules/swap/pure/EthFlow/EthFlowStepper/index.cosmos.tsx:
```diff
@@ -350,6 +313,47 @@ const STEPS: Step[] = [
+      },
+      creation: { ...defaultProps.creation, failed: false },
+      cancellation: { hash: TX, failed: true },
+      refund: { failed: false },
```
you have me lost with this state. Why does it say "xDAI Refunded?"
If I read the state, the order is still OPEN

<img width="1641" alt="image" src="https://user-images.githubusercontent.com/2352112/209008884-3e669c0b-506f-4cb0-b7fc-000a978c8683.png">

alfetopito replied:
`refund: {failed: `false`}`

That means the refund succeeded. Maybe I need to update the state to `cancelled` or `expired`

alfetopito replied:
Or make it so the cancellation failed and the order is filled

anxolin commented on src/cow-react/modules/swap/pure/EthFlow/EthFlowStepper/index.cosmos.tsx:
```diff
@@ -350,6 +313,47 @@ const STEPS: Step[] = [
+        isExpired: false,
+      },
+      creation: { ...defaultProps.creation, failed: false },
+      refund: { hash: TX, failed: true },
```
This state was added, and could happen, but I don't think is well represented. Why if the refund failed we show that it is refunded.

<img width="1314" alt="image" src="https://user-images.githubusercontent.com/2352112/209009594-18c859be-4b77-4c83-8722-7033d4ba2d37.png">

alfetopito replied:
Yeah, I didn't really handled this because I don't know what should happen in that case.
Is it possible the refund tx from the backend to fail?

elena-zh reviewed (approved):
Hey @alfetopito , 1st case LGTM!
2nd case is reported in #1787 

Thanks!
